core/src/main/java/io/questdb/cairo/sql/PageFrameAddressCache.java

```diff
@@ -0,0 +1,177 @@
+            cacheSize += frameAuxPageAddresses.capacity();
+            auxPageSizes.add(frameAuxPageSizes);
+            cacheSize += frameAuxPageSizes.capacity();
+        } else {
+            pageAddresses.add(null);
+            pageSizes.add(null);
+            auxPageAddresses.add(null);
+            auxPageSizes.add(null);
+        }
+
+        frameSizes.add(frame.getPartitionHi() - frame.getPartitionLo());
+        frameFormats.add(frame.getFormat());
+        rowIdOffsets.add(Rows.toRowID(frame.getPartitionIndex(), frame.getPartitionLo()));
+    }
+
+    @Override
+    public void clear() {
+        frameSizes.clear();
+        frameFormats.clear();
+        pageAddresses.clear();
+        auxPageAddresses.clear();
+        pageSizes.clear();
+        auxPageSizes.clear();
+        rowIdOffsets.clear();
+        if (cacheSize < nativeCacheSizeThreshold) {
+            longListPool.clear();
+        } else {
+            longListPool.resetCapacity();
+        }
+        cacheSize = 0;
+    }
+
+    public LongList getAuxPageAddresses(int frameIndex) {
+        return auxPageAddresses.getQuick(frameIndex);
+    }
+
+    public LongList getAuxPageSizes(int frameIndex) {
+        return auxPageSizes.getQuick(frameIndex);
+    }
+
+    public int getColumnCount() {
+        return columnCount;
+    }
+
+    public IntList getColumnTypes() {
+        return columnTypes;
+    }
+
+    public byte getFrameFormat(int frameIndex) {
+        return frameFormats.getQuick(frameIndex);
+    }
+
+    public long getFrameSize(int frameIndex) {
+        return frameSizes.getQuick(frameIndex);
+    }
+
+    public LongList getPageAddresses(int frameIndex) {
+        return pageAddresses.getQuick(frameIndex);
+    }
+
+    public LongList getPageSizes(int frameIndex) {
+        return pageSizes.getQuick(frameIndex);
+    }
+
+    public long getRowIdOffset(int frameIndex) {
+        return rowIdOffsets.getQuick(frameIndex);
+    }
+
+    public boolean hasColumnTops(int frameIndex) {
+        final byte frameFormat = frameFormats.getQuick(frameIndex);
+        assert frameFormat == PageFrame.NATIVE_FORMAT;
+        for (int columnIndex = 0; columnIndex < columnCount; columnIndex++) {
+            if (pageAddresses.getQuick(frameIndex).getQuick(columnIndex) == 0
+                    // VARCHAR column that contains short strings will have zero data vector,
+                    // so for such columns we also need to check that the aux (index) vector is zero.
+                    && auxPageAddresses.getQuick(frameIndex).getQuick(columnIndex) == 0) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean isVarSizeColumn(int columnIndex) {
+        return ColumnType.isVarSize(columnTypes.getQuick(columnIndex));
+    }
+
+    public void of(@Transient RecordMetadata metadata) {
+        columnCount = metadata.getColumnCount();
+        columnTypes.clear();
```

Comment:
perhaps `this.clear()` is needed here?

Reply:
It's not a must as the caller code makes sure to clear the cache on the query execution end, but on the other hand, this call won't hurt. Added it in 5a553229e50cd1d14e4111a4b2e5d3a17de76b1d